aiosmtpd/docs/proxyprotocol.rst

```diff
@@ -0,0 +1,426 @@
+.. _ProxyProtocol:
+
+========================
+ PROXY Protocol Support
+========================
+
+When put behind a "proxy" / load balancer,
+server programs can no longer "see" the original client's actual IP Address and Port.
+
+This also affects ``aiosmtpd``.
+
+The |HAProxyDevelopers|_ have created a protocol called "PROXY Protocol"
+designed to solve this issue.
+You can read the reasoning behind this in `their blog`_.
+
+.. _`HAProxyDevelopers`: https://www.haproxy.com/company/about-us/
+.. |HAProxyDevelopers| replace:: **HAProxy Developers**
+.. _their blog: https://www.haproxy.com/blog/haproxy/proxy-protocol/
+
+This initiative has been accepted and supported by many important software and services
+such as `Amazon Web Services`_, `haproxy`_, `nginx`_, `stunnel`_, `varnish`_, and many others.
+
+.. _Amazon Web Services: https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/enable-proxy-protocol.html
+.. _haproxy: http://cbonte.github.io/haproxy-dconv/2.3/configuration.html#5.2-send-proxy
+.. _nginx: https://nginx.org/en/docs/stream/ngx_stream_proxy_module.html#proxy_protocol
+.. _stunnel: https://www.stunnel.org/static/stunnel.html#proxy
+.. _varnish: https://info.varnish-software.com/blog/proxy-protocol-original-value-client-identity
+
+``aiosmtpd`` implements the PROXY Protocol as defined in |HAProxy2.3.0|_;
```

Comment:
Well, I'm using the commit that has been tagged as `v2.3.0`: https://github.com/haproxy/haproxy/tree/v2.3.0

Unless there is plan to shift the tags around, IMHO referring to it as "version 2.3.0" is not wrong.

Reply:
It is not wrong, but also not right. The PROXY spec evolves independently from HAProxy. Two examples:

- When I implemented the unique id TLV I first sent a patch that simply updated the specification and nothing else. Only afterwards I actually implemented client and server support in HAProxy. Sure, this all happened within a single HAProxy release, but that would not have been a requirement.
- Some future extensions might not be useful in context of HAProxy (e.g. hypothetical stuff related to UDP support). They will be part of the PROXY spec, but not be implemented in HAProxy and instead only in other software (e.g. Amazon's LBs).

Thus referring to the PROXY spec compatibility by the HAProxy branch it appeared in might be misleading. That's why I suggested to use the date instead.

Reply:
I much prefer linking to something with a tag, especially a tag that is production level.

The reasoning being that when something is tagged with a "production" tag, it's frozen.

A commit without a tag is, from my point of view, something "ephemeral and may change". Kind of like my commits in this PR; a certain commit may specify such & such behavior, then several commits later I may change my mind and specify a different behavior as I find a drawback to the previous.

This might be subjective, but that's my stand.